File: DragonsGenerator.API/Models/Spell.cs
using System.Text.Json;

namespace DragonsGenerator.API.Models;

public record Spell(
    //string Guid,
    string Id,
    string Name,
    int Level,
    string School,
    CastingTime CastingTime,
    SpellRange Range,
    SpellDuration Duration,
    SpellComponents Components,
    bool IsRitual,
    bool IsConcentration,
    bool IsCorrupted,
    string Description,
    List<ModularOption> ModularOptions,
    List<string> Classes,
    string? HigherLevels = null
);

public record CastingTime(JsonElement? Amount, string? Unit);

public record SpellRange(JsonElement? Amount, string? Unit);

public record SpellDuration(JsonElement? Amount, string? Unit);

public record SpellComponents(bool V, bool S, string? M);

public record ModularOption(string Name, string Description);

File: DragonsGenerator.WEB/src/app/core/models/Spells/spell.ts
// core/models/Spells/spell.ts

export interface Spell {
  id: string;
  name: string;
  level: number;
  school: string;
  castingTime: CastingTime;
  range: SpellRange;
  duration: SpellDuration;
  components: SpellComponents;
  isRitual: boolean;
  isConcentration: boolean;
  isCorrupted: boolean;
  description: string;
  modularOptions: ModularOption[];
  classes: string[];
  higherLevels?: string | null;
}

export interface CastingTime {
  amount: number | string | null;
  unit: string | null;
}

export interface SpellRange {
  amount: number | string | null;
  unit: string | null;
}

export interface SpellDuration {
  amount: number | string | null;
  unit: string | null;
}

export interface SpellComponents {
  v: boolean;
  s: boolean;
  m: string | null;
}

export interface ModularOption {
  name: string;
  description: string;
}
